jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -430,89 +435,133 @@ private void checkImageInTargetRegistry() {
+                                          results
+                                              .builtImagesAndContainerConfigurationPushResults
+                                              .get()
+                                              .get(builtImage))
                                       .get()
-                                      .get(results.builtImages.get().get(0)))
-                              .get(),
-                          results.builtImages.get().get(0).get(),
-                          results.manifestCheckResult.get().isPresent()));
-              realizeFutures(manifestPushResults);
-              return manifestPushResults.isEmpty()
-                  ? new BuildResult(
-                      results.manifestCheckResult.get().get().getDigest(),
-                      Verify.verifyNotNull(
-                              results
-                                  .builtImagesAndContainerConfigurationPushResults
-                                  .get()
-                                  .get(results.builtImages.get().get(0)))
-                          .get()
-                          .getDigest())
-                  // Manifest pushers return the same BuildResult.
-                  : manifestPushResults.get(0).get();
+                                      .getDigest())
+                              // Manifest pushers return the same BuildResult.
+                              : manifestPushResults.get(0).get();
+                        });
+
+                buildResults.add(buildResult);
+              }
+              return buildResults;
             });
   }
 
   private void loadDocker(DockerClient dockerClient) {
     ProgressEventDispatcher.Factory childProgressDispatcherFactory =
         Verify.verifyNotNull(rootProgressDispatcher).newChildProducer();
 
-    results.buildResult =
+    results.buildResults =
         executorService.submit(
-            () ->
-                new LoadDockerStep(
-                        buildContext,
-                        childProgressDispatcherFactory,
-                        dockerClient,
-                        results.builtImages.get().get(0).get())
-                    .call());
+            () -> {
+              Verify.verify(
+                  results.builtImagesAndBaseImages.get().size() == 1,
+                  "multi-platform image building not supported when pushing to Docker engine");
+              Future<Image> builtImage =
+                  results.builtImagesAndBaseImages.get().keySet().iterator().next();
+
+              return Collections.singletonList(
+                  executorService.submit(
+                      () ->
+                          new LoadDockerStep(
+                                  buildContext,
+                                  childProgressDispatcherFactory,
+                                  dockerClient,
+                                  builtImage.get())
+                              .call()));
+            });
```

Comment:
I think we can block until we get the built `Image`?

```java
              Image builtImage =
                  results.builtImagesAndBaseImages.get().keySet().iterator().next().get();
```

Next, I think we should avoid scheduling another thread. We can use `Futures.immediateFuture` (a utility method to directly instantiate a `Future` from a concrete value) that just wraps a value instead of actually scheduling a thread.

```java
              Image builtImage =
                  results.builtImagesAndBaseImages.get().keySet().iterator().next().get();

              return Collections.singletonList(
                  Futures.immediateFuture(
                      new LoadDockerStep(
                              buildContext,
                              childProgressDispatcherFactory,
                              dockerClient,
                              builtImage)
                          .call()));
```

Reply:
And I think this is more readable:
```java
              BuildResult buildResult =
                  new LoadDockerStep(
                          buildContext, childProgressDispatcherFactory, dockerClient, builtImage)
                      .call();
              return Collections.singletonList(Futures.immediateFuture(buildResult));
```